compiler/rustc_hir_analysis/src/outlives/implicit_infer.rs

```diff
@@ -83,13 +83,34 @@ pub(super) fn infer_predicates(
                 .get(&item_did.to_def_id())
                 .map_or(0, |p| p.as_ref().skip_binder().len());
             if item_required_predicates.len() > item_predicates_len {
-                predicates_added = true;
+                predicates_added.get_or_insert_default().push(item_did);
                 global_inferred_outlives
                     .insert(item_did.to_def_id(), ty::EarlyBinder::bind(item_required_predicates));
             }
         }
 
-        if !predicates_added {
+        if let Some(ids) = predicates_added {
+            if !tcx.recursion_limit().value_within_limit(i) {
+                let msg = if let &[id] = &ids[..] {
+                    format!(
+                        "overflow computing implied lifetime bounds for `{}`",
+                        tcx.def_path_str(id),
+                    )
+                } else {
+                    "overflow computing implied lifetime bounds".to_string()
+                };
+                tcx.dcx()
+                    .struct_span_err(
+                        ids.iter().map(|id| tcx.def_span(*id)).collect::<Vec<_>>(),
+                        msg,
+                    )
+                    .emit();
+                for id in ids {
+                    global_inferred_outlives.shift_remove(&id.to_def_id());
```

Comment:
that seems suboptimal :thinking: what happens if u don't remove the added implied bounds here?

Reply:
hangs in both the old and need solvers due to RECURSION_LIMIT deep types having to be normalized

Reply:
makes sense, we get very deep assoc items. I feel like we can also have cycles of length 2 in which case we still get a hang as we don't remove the implied bounds for the other cycle participant? :thinking: 

vibes, but maybe just fatal error? :grin: